File: 08/solution.ml
open Core
open Stdio

let read_input () = In_channel.read_lines "08/input.txt"

type vec3 = { x : int; y : int; z : int }

let parse_vec3 s =
  match String.split s ~on:',' |> List.map ~f:Int.of_string with
  | [ x; y; z ] -> { x; y; z }
  | _ -> failwith "bad input"

let dist a b =
  let dx = Float.of_int (a.x - b.x) in
  let dy = Float.of_int (a.y - b.y) in
  let dz = Float.of_int (a.z - b.z) in
  Float.sqrt ((dx *. dx) +. (dy *. dy) +. (dz *. dz))

let parse_points lines = List.map lines ~f:parse_vec3 |> List.to_array
let make_uf points = Array.map points ~f:Union_find.create

let sorted_pairs points =
  let n = Array.length points in
  List.init n ~f:(fun i ->
      List.init
        (n - i - 1)
        ~f:(fun j ->
          let k = i + j + 1 in
          (dist points.(i) points.(k), i, k)))
  |> List.concat
  |> List.sort ~compare:(fun (d1, _, _) (d2, _, _) -> Float.compare d1 d2)

let group_sizes uf_nodes =
  let groups = Hashtbl.create (module Int) in
  Array.iteri uf_nodes ~f:(fun _ node ->
      let root_idx =
        Array.findi_exn uf_nodes ~f:(fun _ other ->
            Union_find.same_class node other)
        |> fst
      in
      Hashtbl.incr groups root_idx);
  Hashtbl.data groups

let part1 lines =
  let points = parse_points lines in
  let uf_nodes = make_uf points in
  let pairs = sorted_pairs points in

  List.take pairs 1000
  |> List.iter ~f:(fun (_, i, k) -> Union_find.union uf_nodes.(i) uf_nodes.(k));

  group_sizes uf_nodes
  |> List.sort ~compare:(fun a b -> Int.compare b a)
  |> Fn.flip List.take 3 |> List.fold ~init:1 ~f:( * )

let part2 lines =
  let points = parse_points lines in
  let uf_nodes = make_uf points in
  let pairs = sorted_pairs points in

  let rec go = function
    | [] -> 0
    | (_, i, k) :: rest ->
        let was_different =
          not (Union_find.same_class uf_nodes.(i) uf_nodes.(k))
        in
        Union_find.union uf_nodes.(i) uf_nodes.(k);
        if was_different && List.length (group_sizes uf_nodes) = 1 then
          points.(i).x * points.(k).x
        else go rest
  in
  go pairs

let () =
  let lines = read_input () in
  printf "Part 1: %d\n" (part1 lines);
  printf "Part 2: %d\n" (part2 lines)

File: 08/solution.ts
import * as mnemonist from "mnemonist";

const input = await Bun.file(`${import.meta.dir}/input.txt`).text();
const lines = input.trim().split("\n");

class UnionFind {
	#parent = new Map<string, string>();
	#groups = new Map<string, Vec3[]>();
	#vecMap = new Map<string, Vec3>(); // key -> original vec3

	constructor() {
		this.#parent = new Map<string, string>();
		this.#groups = new Map<string, Vec3[]>();
		this.#vecMap = new Map<string, Vec3>();
	}

	#key(v: Vec3): string {
		return `${v.x},${v.y},${v.z}`;
	}

	find(v: Vec3): string {
		const k = this.#key(v);
		if (!this.#parent.has(k)) {
			this.#parent.set(k, k);
			this.#groups.set(k, [v]);
			this.#vecMap.set(k, v);
		}
		if (this.#parent.get(k) !== k) {
			this.#parent.set(k, this.find(this.#vecMap.get(this.#parent.get(k)!)!));
		}
		return this.#parent.get(k)!;
	}

	union(a: Vec3, b: Vec3): void {
		const rootA = this.find(a);
		const rootB = this.find(b);
		if (rootA !== rootB) {
			this.#parent.set(rootB, rootA);
			this.#groups.get(rootA)!.push(...this.#groups.get(rootB)!);
			this.#groups.delete(rootB);
		}
	}

	getGroup(v: Vec3): Vec3[] {
		return this.#groups.get(this.find(v))!;
	}

	getAllGroups(): Vec3[][] {
		return [...this.#groups.values()];
	}
}

type Vec3 = { x: number; y: number; z: number };
type Node = { dist: number; pair: Vec3[] };

function dist(a: Vec3, b: Vec3) {
	return Math.sqrt((a.x - b.x) ** 2 + (a.y - b.y) ** 2 + (a.z - b.z) ** 2);
}

function toVec3(arr: number[]) {
	if (arr.length !== 3) throw new Error("bad input");
	return { x: arr[0], y: arr[1], z: arr[2] };
}

// make the 1000 (test: 10) shortest pair-wise connections and count circuit groups
// build min heap on pair-wise distances and extract min
// then implement a dynamic union-find to build groups
function part1() {
	const nsmallest = 1000;
	const points = lines.map((l) => l.split(",").map(Number)).map(toVec3);
	const heap = new mnemonist.MinHeap<Node>((a, b) => a.dist - b.dist);
	const uf = new UnionFind();

	for (const p of points) {
		uf.find(p);
	}

	// all pairs
	for (let i = 0; i < points.length; i++) {
		for (let k = i + 1; k < points.length; k++) {
			const [a, b] = [points[i], points[k]];
			const d = dist(a, b);
			heap.push({ dist: d, pair: [a, b] });
		}
	}

	for (let i = 0; i < nsmallest; i++) {
		const node = heap.pop()!;
		const [a, b] = node.pair;
		uf.union(a, b);
	}

	const largest3 = uf
		.getAllGroups()
		.sort((a, b) => b.length - a.length)
		.slice(0, 3);

	return largest3.reduce((acc, g) => acc * g.length, 1);
}

// find when the MST (minimum spanning tree) becomes fully connected
function part2() {
	const points = lines.map((l) => l.split(",").map(Number)).map(toVec3);
	const heap = new mnemonist.MinHeap<Node>((a, b) => a.dist - b.dist);
	const uf = new UnionFind();

	for (const p of points) {
		uf.find(p);
	}

	// build heap with all pairs
	for (let i = 0; i < points.length; i++) {
		for (let k = i + 1; k < points.length; k++) {
			const [a, b] = [points[i], points[k]];
			const d = dist(a, b);
			heap.push({ dist: d, pair: [a, b] });
		}
	}

	while (uf.getAllGroups().length > 1) {
		const node = heap.pop()!;
		const [a, b] = node.pair;

		if (uf.find(a) !== uf.find(b)) {
			uf.union(a, b);

			// last pair to create 1 circuit
			if (uf.getAllGroups().length === 1) {
				return a.x * b.x;
			}
		}
	}

	return 0;
}

console.log("Part 1:", part1());
console.log("Part 2:", part2());
